pubsub/kafka/kafka.go

```diff
@@ -24,32 +26,50 @@ import (
 
 type PubSub struct {
 	kafka  *kafka.Kafka
-	topics map[string]bool
+	topics map[string]pubsub.Handler
+	lock   sync.RWMutex
 }
 
 func (p *PubSub) Init(metadata pubsub.Metadata) error {
-	p.topics = make(map[string]bool)
 	return p.kafka.Init(metadata.Properties)
 }
 
 func (p *PubSub) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handler) error {
-	topics := p.addTopic(req.Topic)
+	// Add the topic to the list of those we subscribe to
+	p.lock.Lock()
+	p.topics[req.Topic] = handler
+	topics := p.topicList()
+	p.lock.Unlock()
 
-	return p.kafka.Subscribe(topics, req.Metadata, newSubscribeAdapter(handler).adapter)
```

Comment:
I suggestion to move the update code to kafka common code, not in kafka pubsub (we have kafka binding and we need to reuse code between kafka components )

Reply:
I moved the routing logic to the common Kafka code.